src/Model/Customer/Address/ValidateAddress.php

```diff
@@ -0,0 +1,126 @@
+use Magento\Customer\Api\Data\AddressInterface;
+use Magento\Customer\Api\Data\AddressInterfaceFactory;
+use Magento\Customer\Api\Data\RegionInterfaceFactory;
+use Magento\Directory\Helper\Data as DirectoryData;
+use Magento\Directory\Model\ResourceModel\Region\CollectionFactory as RegionCollectionFactory;
+use Magento\Framework\GraphQl\Exception\GraphQlInputException;
+
+class ValidateAddress extends SourceValidateAddress
+{
+    /**
+     * @var AddressInterfaceFactory
+     */
+    private $addressFactory;
+
+    /**
+     * @var RegionInterfaceFactory
+     */
+    private $regionFactory;
+
+    /**
+     * @var DirectoryData
+     */
+    protected $directoryData;
+
+    /**
+     * @var RegionCollectionFactory
+     */
+    protected $regionCollectionFactory;
+
+    /**
+     * @var ExtractCustomerAddressData
+     */
+    protected $extractCustomerAddressData;
+
+    /**
+     * ValidateCustomerData constructor.
+     *
+     * @param AddressInterfaceFactory $addressFactory
+     * @param RegionInterfaceFactory $regionFactory
+     * @param DirectoryData $directoryData
+     * @param RegionCollectionFactory $regionCollectionFactory
+     * @param ExtractCustomerAddressData $extractCustomerAddressData
+     */
+    public function __construct(
+        AddressInterfaceFactory $addressFactory,
+        RegionInterfaceFactory $regionFactory,
+        DirectoryData $directoryData,
+        RegionCollectionFactory $regionCollectionFactory,
+        ExtractCustomerAddressData $extractCustomerAddressData
+    ) {
+        parent::__construct(
+            $addressFactory,
+            $regionFactory,
+            $directoryData,
+            $regionCollectionFactory,
+            $extractCustomerAddressData
+        );
+
+        $this->extractCustomerAddressData = $extractCustomerAddressData;
+        $this->regionCollectionFactory = $regionCollectionFactory;
+        $this->directoryData = $directoryData;
+    }
+
+    /**
+     * Validate customer address data
+     *
+     * @param AddressInterface $address
+     * @throws GraphQlInputException
+     */
+    public function execute(AddressInterface $address): void
+    {
+        $addressData = $this->extractCustomerAddressData->execute($address);
+
+        if (isset($addressData['country_code'])) {
+            $isRegionRequired = $this->directoryData->isRegionRequired($addressData['country_code']);
+
+            if ($isRegionRequired && is_null($addressData['region']['region_id'])) {
+                throw new GraphQlInputException(__('A region_id is required for the specified country code'));
+            }
+            $regionCollection = $this->regionCollectionFactory
+                ->create()
+                ->addCountryFilter($addressData['country_code']);
+
+            if ($isRegionRequired) {
+                if (!isset($addressData['region']['region_code'])) {
+                    $regionCollection->addRegionCodeFilter($addressData['region']['region_code']);
```

Comment:
If you `return` here, the `else` block becomes unnecessary and can be elevated one scope up.